python/lsst/daf/butler/script/rewriteSqliteRegistry.py

```diff
@@ -0,0 +1,202 @@
+    del config["registry", "db"]
+
+    # Force the new temporary butler repo to have an explicit path
+    # to the existing datastore.  This will only work for a FileDatastore
+    # so try that and if it doesn't work warn but continue since either
+    # this is a new typoe of datastore or a chained datastore of some kind.
+    # For now only handle the simple case.
+    # NOTE: Execution Butler creation has a similar problem with working
+    # out how to refer back to the original datastore.
+    try:
+        config["datastore", "root"] = str(source_butler.datastore.root)  # type: ignore
+
+        # Force the name of the datastore since that should not
+        # change from the source (and will if set from root)
+        config["datastore", "name"] = source_butler.datastore.name
+    except AttributeError:
+        log.warning("Migration is designed for FileDatastore yet something different was encountered."
+                    " Attempting migration anyhow. It should work so long as <butlerRoot> is not used"
+                    " in config.")
+
+    # Export all the required information from the source butler
+    exported = export_non_datasets(source_butler)
+
+    # Create a new butler with this config as the seed but ensuring that it
+    # does not overwrite the datastore root.
+    dest_config = Butler.makeRepo(dest, config,
+                                  dimensionConfig=source_butler.registry.dimensions.dimensionConfig,
+                                  forceConfigRoot=False)
+
+    # Create destination butler
+    dest_butler = Butler(dest_config, writeable=True)
+
+    # Import the data to the new butler
+    import_non_datasets(dest_butler, exported)
+
+    # Now transfer all the datasets (removing duplicates but needing to
+    # know the order).
+    source_refs = list(set(source_butler.registry.queryDatasets(..., collections=...)))
+
+    # If this is int to UUID we force "raw" to generate reproductible UUID.
+    dest_refs = dest_butler.transfer_from(source_butler, source_refs, skip_missing=False,
+                                          id_gen_map={"raw": DatasetIdGenEnum.DATAID_TYPE})
+
+    # Map source ID to destination ID.
+    source_to_dest = {source.getCheckedId(): dest for source, dest in zip(source_refs, dest_refs)}
+
+    # Create any dataset associations
+    create_associations(source_butler, dest_butler, source_to_dest)
+
+    # The registry is now rewritten but the question is what to do next.
+    # This new registry points to the original datastore so people shouldn't
+    # think they can delete the original path completely.
+    # Do we:
+    # 1) Tell the user to move this new registry into the original location?
+    # 2) Move the original to a backup and move the new one to original,
+    #    deleting the destination? Copy it instead of move?
+    # 3) Not even ask for a destination location but have a temporary
+    #    location that we clean up after moving the new registry to old place.
```

Comment:
If the intent is to replace original one then renaming original to a backup and new one to original is probably OK.